File: go/sieve.go
package main

import (
	"math"
)

func Sieve(size int) []int {
	if size < 2 {
		return []int{}
	}

	a := make([]int, size/2)
	q := int(math.Sqrt(float64(size)))
	factor := 3

	for factor < q {
		for i := factor; i < size; i += 2 {
			if a[i>>1] == 0 {
				factor = i
				break
			}
		}

		for i := factor * factor; i < size; i += factor * 2 {
			a[i>>1] = 1
		}

		factor += 2
	}

	a[0] = 2
	prime_idx := 1
	for i := 3; i < size; i += 2 {
		if a[i>>1] == 0 {
			//res = append(res, i)
			a[prime_idx] = i
			prime_idx += 1
		}
	}

  return a[0:prime_idx]
}

File: py3/sieve.py
from typing import List
import math


def sieve(size: int) -> List[int]:
    if size < 2:
        return []

    # keep track of odd values only (don't look at even indices)
    size2 = size // 2;
    a = bytearray(size2)
    q = math.floor(math.sqrt(size))
    factor = 3

    while factor < q:
        for num in range(factor//2, size2, 1):
            if a[num] == 0:
                # found a new prime
                factor = num*2 + 1
                break

        # remove all multiples of this prime
        for num in range((factor * factor)//2, size2, factor):
            a[num] = 1

        factor += 2

    return [2] + [i for i in range(3, size, 2) if a[i//2] == 0]
